scipy/signal/_peak_finding.py

```diff
@@ -336,78 +345,62 @@ def peak_prominences(x, peaks, wlen=None):
+    except TypeError:
+        if peaks.size == 0:
+            # Empty arrays default to np.float64 but are valid input
+            peaks = np.array([], dtype=np.intp)
+        else:
+            raise TypeError("Cannot safely cast `peaks` to dtype('intp')")
     if peaks.ndim != 1:
         raise ValueError('`peaks` must have exactly one dimension')
-    if x.size <= peaks.max():
-        raise ValueError('an index in `peaks` exceeds the size of `x`')
-    if not np.issubdtype(peaks.dtype, np.integer):
-        raise ValueError('`peaks` must be an array of integers')
-    if wlen is not None and wlen < 3:
-        raise ValueError('`wlen` must be at least 3')
-
-    # Prepare return arguments
-    prominences = np.zeros(peaks.size)
-    left_bases = np.zeros(peaks.size, dtype=int)
-    right_bases = np.zeros(peaks.size, dtype=int)
-
-    for i, peak in enumerate(peaks):
-        # If wlen is twice the size of x the symmetric window always covers
-        # the full x
-        if wlen is not None and wlen < x.size * 2:
-            wlen = int(wlen)
-            # Calculate window borders around the evaluated peak
-            wleft = peak - wlen // 2
-            wright = peak + wlen // 2
-            # Handle border cases
-            wleft = 0 if wleft < 0 else wleft
-            wright = x.size if x.size < wright else wright
-            # Use slice for prominence calculation
-            window = x[wleft:wright]
-            # Correct peak position in x
-            peak -= wleft
-        else:
-            # Use full x for prominence calculation
-            window = x
-            wleft = 0
 
-        # Positions where window is larger than current peak height
-        greater_peak = np.where(window > window[peak])[0]
-
-        try:
-            # Nearest position to the left of peak with
-            # window[left] > window[peak]
-            left = greater_peak[greater_peak < peak].max()
-        except ValueError:
-            left = 0
-        try:
-            # Nearest position to right of peak with
-            # window[right] > window[peak]
-            right = greater_peak[greater_peak > peak].min()
-        except ValueError:
-            right = None
-
-        # Base indices to the left and right of peak in window
-        left_bases[i] = window[left:peak].argmin() + left
-        right_bases[i] = window[peak:right].argmin() + peak
-
-        # Calculate lowest contour and its vertical distance to peak
-        lowest_contour = max(window[left_bases[i]], window[right_bases[i]])
-        prominences[i] = window[peak] - lowest_contour
-
-        # Correct window offset
-        left_bases[i] += wleft
-        right_bases[i] += wleft
+    if wlen is None:
+        wlen = -1  # Inner function expects int -> None == -1
+    elif 1 < wlen:
+        # Round up to next positive integer; rounding up to next odd integer
+        # happens implicitly inside the inner function
+        wlen = math.ceil(wlen)
```

Comment:
You likely want `int(math.ceil(wlen))` here.

Reply:
Seems like `math.ceil` should always return an integral value, at least that's what the [docs](https://docs.python.org/3.6/library/math.html#math.ceil) say. So I didn't see the need.
```
>>> type(math.ceil(3.4))
int
>>> type(math.ceil(np.array(2.3)))
int
```

Reply:
Not on 2.7

https://docs.python.org/2.7/library/math.html#math.ceil

Reply:
Oh, good catch. But that doesn't seem to be a problem because the unit tests pass even if wlen is a float under 2.7. Maybe that's because of [automatic type conversion](http://docs.cython.org/en/latest/src/reference/language_basics.html#automatic-type-conversion) when passing wlen to the Cython function?

Should we rely on this?

Reply:
It's probably better to be explicit about it.